### PR TITLE
if configuration is defined programmatically load it and ignore others

### PR DESCRIPTION
**Pull request by vadimkim**
Fix hotrod-client.properties overwrite for the configuration defined by program

### CODE REVIEW

File: infinispan-spring-boot-starter-remote/src/main/java/org/infinispan/spring/starter/remote/InfinispanRemoteAutoConfiguration.java [hunk 1]
```diff
@@ -64,10 +64,14 @@ public RemoteCacheManager remoteCacheManager() throws IOException {
...
       boolean hasHotRodPropertiesFile = ctx.getResource(infinispanProperties.getClientProperties()).exists();
       boolean hasConfigurer = infinispanRemoteConfigurer != null;
+      boolean hasConfiguration = infinispanConfiguration != null;
       boolean hasProperties = StringUtils.hasText(infinispanProperties.getServerList());
 
       org.infinispan.client.hotrod.configuration.Configuration configuration;
-      if (hasConfigurer) {
+      
+      if (hasConfiguration) {
+         configuration = infinispanConfiguration;
```
Comment (slaskawi) [inline infinispan-spring-boot-starter-remote/src/main/java/org/infinispan/spring/starter/remote/InfinispanRemoteAutoConfiguration.java]: This change looks fine although it it's not symmetric with `InfinispanEmbeddedAutoConfiguration`. The `InfinispanEmbeddedAutoConfiguration` checks whether an XML file is accessible and then checks other configuration options. I believe this is a correct approach where configuration files are somewhat "stronger" than POJO configuration.

May I ask you to move `configuration = infinispanConfiguration` into if statement below (where it was before)?

Comment (slaskawi) [reply infinispan-spring-boot-starter-remote/src/main/java/org/infinispan/spring/starter/remote/InfinispanRemoteAutoConfiguration.java]: Alternatively to this, we can grab POJO configuration and merge properties and configurer configurations into it using `ConfigurationBuilder#read()` method but I think this kind of solution is less intuitive.

**Comment by vadimkim** [comment]
It seems existing source code at master branch already contains condition.
`...else if (infinispanConfiguration != null)..`
while 2.0.0.Alpha1 release binary - doesn't.  I think you can reject my pull request and release existing code for the problem to be solved

**Comment by slaskawi** [comment]
Ok, I've just released `2.0.0.Beta1` (https://github.com/infinispan/infinispan-spring-boot/releases/tag/2.0.0.Beta1). Could you please retry with the latest release?

**Comment by vadimkim** [comment]
I have tested and code fails because condition 
`if (this.infinispanConfiguration == null) ` is behind **else** that is hit only if no other configuration options are present. Particularly:
```
  } else {
            if (this.infinispanConfiguration == null) {
                throw new IllegalStateException("Not enough data to create RemoteCacheManager. Check InfinispanRemoteCacheManagerCheckerand update conditions.");
            }

            builder = (new ConfigurationBuilder()).read(this.infinispanConfiguration);
            this.cacheCustomizers.forEach((c) -> {
                c.customize(builder);
            });
            configuration = builder.build();
        }
```
Do I uderstand correctly that goals are:
1. Read configuration files first and create configuration
2. If there is a no configuration files load `@Bean` from the code
3. If there is no configuration files and no bean definition - rise IllegalStateException "Not enough data..."
4. If there are both configuration files and Bean -- merge them together and file properties go first? 

I can propose implementation of p.4 if my understanding of the task is correct.

**Comment by slaskawi** [comment]
@vadimkim You're somewhat correct but let me make this a bit clearer...

At first we check if your code contains a `Configurer`. Something like this:

```
@Bean
   public InfinispanRemoteConfigurer infinispanRemoteConfigurer() {
      return () -> new ConfigurationBuilder().addServer().host("127.0.0.1").port(PORT).build();
   }
```

If it does, we use it. 

Next, we check if your classpath contains `hotrod-client.properties`. It should contain all properties in our good old format described in Infinispan manual. You may find an example [here](https://github.com/infinispan/infinispan-spring-boot/blob/master/infinispan-spring-boot-starter-remote/src/test/resources/test-hotrod-client.properties).

Next, we check if your classpath contains Spring-like properties. All properties in Spring Boot are typically stored in `application.properties` or `application.yaml` (you may read more about it in the [manual](https://docs.spring.io/spring-boot/docs/current/reference/html/howto-properties-and-configuration.html)). All Infinispan properties start with `infinispan.remote`, here's [an example](https://github.com/infinispan/infinispan-spring-boot/blob/master/infinispan-spring-boot-starter-remote/src/test/resources/test-application.properties).

Finally, in our last try, we check if there's a `@bean` with Hot Rod configuration (`org.infinispan.client.hotrod.configuration.Configuration`). 

If all of the above fail, we throw an error.

**Comment by vadimkim** [comment]
@slaskawi the problem is that properties are not equal. For example: hotrod-client.properties file can't contain security and keystore definition. There are no such options according to documentation. `@bean` with Hot Rod configuration allow to customize everything for the hotrod client. 
I didn't analyze all possible variants but basically we are compare configurations that are not equal. The questions are:
1. Can InfinispanRemoteConfigurer define all client properties including security yes/no?
2. Can hotrod-client.properties define all client properties -- the answer is no
3. Can spring boot application YAML define all client properties including security yes/no?
4. Can hotrod configuration bean define all  client properties -- the answer is yes

**Comment by slaskawi** [comment]
@vadimkim Answering specific questions:

> Can InfinispanRemoteConfigurer define all client properties including security yes/no?

Yes it can. 

```
@Bean
   public InfinispanRemoteConfigurer infinispanRemoteConfigurer() {
      return () -> new ConfigurationBuilder().security()...
   }
```

> Can hotrod-client.properties define all client properties -- the answer is no

As @tristantarrant pointed out, yes it can. Unfortunately some of them were undocumented... but that's another issue.

> Can spring boot application YAML define all client properties including security yes/no?

No. I created https://github.com/infinispan/infinispan-spring-boot/issues/47 to address this.

> Can hotrod configuration bean define all client properties -- the answer is yes

Yes.

**Comment by vadimkim** [comment]
to quickly sum up:  we shouldn't change existing code, but rather document all hotrod-client.properties and add configuration options to srping boot application. Then everything will work smoothly. Right?

P.S. @tristantarrant  I was trying to create hotrod-client.properties and stuck with 
`infinispan.client.hotrod.auth_callback_handler = ee.demo.infinispan.security.TestCallHandler`
Exception:
`nested exception is org.infinispan.commons.CacheConfigurationException: Unable to instantiate class ee.demo.infinispan.security.TestCallHandler`
Without callback class application thows transport exception

**Comment by tristantarrant** [comment]
A manually defined callback handler is not necessary since Infinispan 9.2 (https://issues.jboss.org/browse/ISPN-8794) when using PLAIN, DIGEST-MD5 and EXTERNAL. Can you paste your stack trace and properties file in a gist so I can verify ?

**Comment by vadimkim** [comment]
@tristantarrant sorry for the late reply and incomplete error log output. I tend not to work on weekends. 
Fresh test shows this:
```
org.infinispan.commons.CacheConfigurationException: java.lang.IllegalArgumentException: File does not contain valid certificates: /home/vadimk/Projects/spring-http-session-infinispan/src/main/resources/truststore.jks
	at org.infinispan.client.hotrod.impl.transport.netty.ChannelInitializer.initSsl(ChannelInitializer.java:136)
	at org.infinispan.client.hotrod.impl.transport.netty.ChannelInitializer.initChannel(ChannelInitializer.java:77)
	at io.netty.channel.ChannelInitializer.initChannel(ChannelInitializer.java:113)
	at io.netty.channel.ChannelInitializer.handlerAdded(ChannelInitializer.java:105)
```
The keystore is the same keystore that I use for manual test and  pure `@Bean` autoconfiguration. Both works.
Corresponding lines at hotrod-client.properties:
```
infinispan.client.hotrod.use_ssl = true
infinispan.client.hotrod.trust_store_path = /home/vadimk/Projects/spring-http-session-infinispan/src/main/resources/truststore.jks
infinispan.client.hotrod.trust_store_type = jks
infinispan.client.hotrod.trust_store_password = storePass
```
I have tried to add/remove "trustore_type", but there were no difference

**Comment by slaskawi** [comment]
The error message seems to be self-descriptive. Something went wrong when you generated your TrustStore. 

I highly recommend you to look into those two examples:

1. Multi-tenancy demo, where we generate both keystore and truststore in one go: https://github.com/infinispan-demos/infinispan-openshift-multitenancy/blob/master/create_certs.sh
2. Our server side test suite where we use Maven plugins to generate those certificates: https://github.com/infinispan/infinispan/blob/master/server/integration/testsuite/pom.xml#L540

**Comment by tristantarrant** [comment]
@vadimkim can you send me the jks file privately please ?

**Comment by vadimkim** [comment]
@slaskawi  I have generated keystore with java keytool manually.
@tristantarrant  sent you private e-mail with keystore to `dataforte.net` account
For all the rest of the methods I tried this keystore worked. I think it could be issue with keystore loader/parser for hotrod native client.

**Comment by vadimkim** [comment]
Yes, it works !

**Comment by slaskawi** [comment]
Seems the problem is solved. Closing this one.

**Pull request closed without merging**